core/include/traccc/utils/subspace.hpp

```diff
@@ -0,0 +1,119 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/definitions/primitives.hpp"
+#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/definitions/track_parametrization.hpp"
+
+// System include(s).
+#include <array>
+
+namespace traccc {
+
+template <typename algebra_t, typename algebra_t::size_type kFullSize,
+          typename algebra_t::size_type kSize = 2u>
+struct subspace {
+    static_assert(1u <= kSize, "Subspace size must be at least 1");
+    static_assert(kSize <= kFullSize,
+                  "Subspace can only be as large as the full space");
+
+    public:
+    // Type declarations
+    using matrix_operator = typename algebra_t::matrix_actor;
+    using size_type = typename matrix_operator::size_ty;
+    template <size_type ROWS, size_type COLS>
+    using matrix_type =
+        typename matrix_operator::template matrix_type<ROWS, COLS>;
+
+    using subspace_vector = matrix_type<kSize, 1u>;
+    using fullspace_vector = matrix_type<kFullSize, 1u>;
+    using projection_matrix = matrix_type<kSize, kFullSize>;
+    using expansion_matrix = matrix_type<kFullSize, kSize>;
+
+    /// Construct from a container of axis indices.
+    ///
+    /// @param indices Unique, ordered indices
+    TRACCC_HOST_DEVICE
+    constexpr subspace(const std::array<size_type, kSize>& indices) {
+        for (size_type i = 0u; i < kSize; ++i) {
+            assert((indices[i] < kFullSize) and
+                   "Axis indices must be within the full space");
+            if (0u < i) {
+                assert((indices[i - 1u] < indices[i]) and
+                       "Axis indices must be unique and ordered");
+            }
+        }
+        for (size_type i = 0; i < kSize; ++i) {
+            m_axes[i] = static_cast<size_type>(indices[i]);
+        }
+    }
+
+    /// Size of the subspace.
+    TRACCC_HOST_DEVICE
+    static constexpr size_type size() { return kSize; }
+    /// Size of the full vector space.
+    TRACCC_HOST_DEVICE
+    static constexpr size_type fullSize() { return kFullSize; }
```

Comment:
Interesting... This is very "modern C\+\+", but in the olden days we would've simply done this with:

```c++
static const size_type size = kSize;
static const size_type fullSize = kFullSize;
```

Well, maybe now using `constexpr` instead of `const`. But mainly, defining these as functions seems to be a bit of an overkill.

Reply:
Fixed